### PR TITLE
ci: unified build/test container

### PR DESCRIPTION
**Pull request by m-ildefons**
- Use unified build/test container for PR testing
- Save results into status repo

This is how the workflow is laid out:

```mermaid
graph TD;
  B1[build unit test container] --> R1[run unit test container];
  R1 -- Failure --> R[Reject]:::red;
  R1 -- Success --> B2[build s3gw container];

  B1 -- Docker Cache --> B2;

  B2 --> R2[run s3gw container & integration tests];
  R2 -- Failure --> R;
  R2 -- Success --> R3[run s3gw container & smoke tests];

  R3 -- Failure --> R;
  R3 -- Success --> R4[run s3gw container & s3tests as regression tests];

  R4 -- Failure --> R;
  R4 -- Success --> S[store s3tests results]:::green;

  classDef red fill:#b21
  classDef green fill:#194
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

### CODE REVIEW

**Review by jecluis — COMMENTED** [review]
lgtm

**Review by jecluis — DISMISSED** [review]
meant as an approval ;)

**Comment by jecluis** [comment]
although it failed, I stand by my comment: it looks good. :P

**Comment by m-ildefons** [comment]
Thanks... the failure was due to a `/` in `${{ github.ref_name }}`, which when expanded into a path will of course break any `mv` invocation, since it is interpreted as subdirectory, which doesn't exist :facepalm: 
I'll ensure it actually works before marking this PR ready ;-)

**Comment by 0xavi0** [comment]
CI is failing because a variable needed to pass the extra parameter to activate the lifecycle debug interval is not exported.

See https://github.com/aquarist-labs/s3gw/pull/475 for more details.

I would merge the PR above before and re-run CI jobs for this one.

**Comment by 0xavi0** [comment]
I see ccache is dropped, was that intentional?

I would store ccache files as soon as build is successful, no matter if tests are passing... otherwise, if we store ccache upon successful CI, we could have an issue in a s3test and will need to rebuild if ccache is dirty, right?

**Comment by m-ildefons** [comment]
One problem with ccache is dirty builds and needing to clean-rebuild, which can be a bit tricky as you'd need to set up some way to signal the CI that "this particular rebuild is special and needs to do something else than usual".
But the main reason is that it provided no real speedup on the self-hosted workers.
The thing is that the ccache directory is several gigabytes in size and the self-hosted workers don't have fast up/download speeds to GitHub. So moving the cache between builds would easily add another ten to twenty minutes to the setup and then another ten to twenty minutes to the cleanup before the workflow result is reported to the PR. With clean rebuild times being around 25 minutes on the self-hosted workers, about half of which is dependency installation, not compilation anyways, the benefit of shoveling the cache around is questionable at best.
And then again, adding self-hosted cache management to make the uploads/downloads faster adds quite a bit of complexity, not just for the pipeline but also operationally. So I figured it just wouldn't be worth the trouble.

**Comment by irq0** [comment]
> The thing is that the ccache directory is several gigabytes in size and the self-hosted workers don't have fast up/download speeds to GitHub. So moving the cache between builds would easily add another ten to twenty minutes to the setup and then another ten to twenty minutes to the cleanup before the workflow result is reported to the PR. With clean rebuild times being around 25 minutes on the self-hosted workers, about half of which is dependency installation, not compilation anyways, the benefit of shoveling the cache around is questionable at best.

Second that. I saw multiple runs where the cache download rate went down to the 100 kbps range. 
https://github.com/mozilla/sccache could be interesting in the future, may even be an opportunity to dog food s3gw

**Review by giubacc — APPROVED** [review]

File: .github/workflows/test-s3gw.yml [hunk 1]
```diff
@@ -8,82 +8,106 @@ on:
...
+          sudo apt-get update
+          sudo apt-get install -y \
+            bc \
+            wget \
+            s3cmd
+
+          # Unfortunately, since yq is only available through snap on Ubuntu and
+          # that doesn't work in docker containers (at least not out of the
+          # box), this abomination is the way to go to install yq.
+          echo "${YQ_CHECKSUM}  yq" >> checksum
```
Comment (irq0) [inline .github/workflows/test-s3gw.yml:75]: nit/optional: bash fd redirection saves the temp file: sha256sum -c <(echo "${YQ_CHECKSUM}  yq")

**Comment by irq0** [comment]
Fixes https://github.com/aquarist-labs/s3gw/issues/479

**Pull request merged**